src/Raven.Server/Documents/Queries/GraphQueryOrderByFieldSorter.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections.Generic;
+using Raven.Client.Json;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries
+{
+    internal class GraphQueryOrderByFieldSorter: IComparer<GraphQueryRunner.Match>
+    {
+        private OrderByField _field;
+        private string _alias;
+        private BlittablePath _path;
+
+        public GraphQueryOrderByFieldSorter(OrderByField field)
+        {
+            var fieldName = field.Name.Value;
+            var indexOfDot = fieldName.IndexOf('.');
+            if (indexOfDot < 0)
+                throw new NotSupportedException($"{GetType().Name} got an order by field: {fieldName} that isn't in the expected format of alias.fieldName");
+            _alias = fieldName.Substring(0, indexOfDot);
+            _path = new BlittablePath(fieldName.Substring(indexOfDot + 1, fieldName.Length - indexOfDot - 1));
+            _field = field;
+        }
+
+
+
+        public int Compare(GraphQueryRunner.Match x, GraphQueryRunner.Match y)
+        {
+            int order = _field.Ascending ? 1 : -1;
+            object xObject = null;
+            object yObject = null;
+            var xResult = x.GetResult(_alias);
+            var yResult = y.GetResult(_alias);
+            switch (xResult)
+            {
+                case Document xDocument:
+                    if (yResult is Document yDocument)
+                    {
+                        xDocument.EnsureMetadata();
+                        yDocument.EnsureMetadata();
+                        xObject = _path.Evaluate(xDocument.Data, true);
+                        yObject = _path.Evaluate(yDocument.Data, true);
+                        break;
+                    }
+                    ThrowMissmatchTypes(xResult, yResult);                   
+                    break;
+                case BlittableJsonReaderObject bjroX:
+                    if (yResult is BlittableJsonReaderObject bjroY)
+                    {
+                        xObject = _path.Evaluate(bjroX, true);
+                        yObject = _path.Evaluate(bjroY, true);
+                        break;
+                    }
+                    ThrowMissmatchTypes(xResult, yResult);
+                    break;
+                case LazyStringValue xLazyStringValue:
+                    return xLazyStringValue.CompareTo(yResult) * order;
+                case string xString:
+                    if (yResult is string yString)
+                    {
+                        xObject = xString;
+                        yObject = yString;
+                    }
+                    ThrowMissmatchTypes(xResult, yResult);
+                    break;
+                case LazyNumberValue xLazyNumberValue:
+                    return xLazyNumberValue.CompareTo(yResult) * order;                
```

Comment:
i expect LazyNumber Value to throw in this case, i was trying to avoid yet another switch because the other side could be LazyNumber int long double i guess there is no avoiding it

Reply:
I believe we must, yes